pkg/kv/kvnemesis/validator.go

```diff
@@ -998,45 +1118,81 @@ func mustGetStringValue(value []byte) string {
 	return string(b)
 }
 
-func validReadTimes(
-	b *pebble.Batch, key roachpb.Key, value []byte, anyValueAccepted bool,
-) disjointTimeSpans {
-	var validTimes disjointTimeSpans
-	end := hlc.MaxTimestamp
-
-	iter := b.NewIter(nil)
+func validReadTimes(b *pebble.Batch, key roachpb.Key, value []byte) disjointTimeSpans {
+	var hist []storage.MVCCValue
+	iter := b.NewIter(&pebble.IterOptions{
+		KeyTypes: storage.IterKeyTypePointsAndRanges,
+	})
 	defer func() { _ = iter.Close() }()
 	iter.SeekGE(storage.EncodeMVCCKey(storage.MVCCKey{Key: key}))
+
 	for ; iter.Valid(); iter.Next() {
 		mvccKey, err := storage.DecodeMVCCKey(iter.Key())
 		if err != nil {
 			panic(err)
 		}
+
+		hasPoint, hasRange := iter.HasPointAndRange()
+		if hasRange && iter.RangeKeyChanged() {
+			k, ek := iter.RangeBounds()
+			sp := roachpb.Span{Key: k, EndKey: ek}
+			if !sp.ContainsKey(key) {
+				// If we see a range key that doesn't even contain the key,
```

Comment:
nit: you could avoid all these checks by setting `LowerBound: key, UpperBound: key.Next()` on the iterator. In that case, all point keys and range keys you encounter are guaranteed to overlap the seek key.